tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,721 @@
+            unsafe { (*self.0).store(s.to_repr(), ordering) }
+        }
     }
+}
 
-    /// The current entry state as known by the timer. This is not the value of
-    /// `state`, but lets the timer know how to converge its state to `state`.
-    pub(crate) fn when_internal(&self) -> Option<u64> {
-        unsafe { *self.when.get() }
-    }
+#[cfg(loom)]
+mod state_cell {
+    use super::EntryState;
+    use crate::loom::sync::atomic::{AtomicU8, Ordering};
 
-    pub(crate) fn set_when_internal(&self, when: Option<u64>) {
-        unsafe {
-            *self.when.get() = when;
+    /// Because loom needs extra tracking around when we start and stop using an UnsafeCell,
+    /// we need to make sure the UnsafeCell isn't destroyed before we finish up our bookkeeping.
+    ///
+    /// As such, when running under loom, we set up an Arc to keep the cell alive.
+    /// (We don't use a loom Arc because this is not part of the production model)
+    pub(super) struct StateCell(std::sync::Arc<AtomicU8>);
+    impl StateCell {
+        pub(super) fn new() -> Self {
+            Self(std::sync::Arc::new(AtomicU8::new(
+                EntryState::NotRegistered.to_repr(),
+            )))
+        }
+
+        pub(super) fn erase(&self) -> StateRef {
+            StateRef(self.0.clone())
+        }
+
+        pub(super) fn get(&self, ordering: Ordering) -> EntryState {
+            unsafe { EntryState::from_repr(self.0.load(ordering)) }
+        }
+
+        pub(super) fn set(&self, s: EntryState, ordering: Ordering) {
+            self.0.store(s.to_repr(), ordering)
         }
     }
 
-    /// Called by `Timer` to load the current value of `state` for processing
-    pub(crate) fn load_state(&self) -> Option<u64> {
-        let state = self.state.load(SeqCst);
+    pub(super) struct StateRef(std::sync::Arc<AtomicU8>);
+
+    impl StateRef {
+        pub(super) fn get(&self, ordering: Ordering) -> EntryState {
+            unsafe { EntryState::from_repr(self.0.load(ordering)) }
+        }
 
-        if is_elapsed(state) {
-            None
-        } else {
-            Some(state)
+        pub(super) fn set(&self, s: EntryState, ordering: Ordering) {
+            self.0.store(s.to_repr(), ordering)
         }
     }
+}
 
-    pub(crate) fn is_elapsed(&self) -> bool {
-        let state = self.state.load(SeqCst);
-        is_elapsed(state)
+use state_cell::StateCell;
+
+impl std::fmt::Debug for StateCell {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "StateCell({:?})", self.get(Ordering::SeqCst))
     }
+}
 
-    pub(crate) fn fire(&self, when: u64) {
-        let mut curr = self.state.load(SeqCst);
+/// A timer entry.
+///
+/// This is the handle to a timer that is controlled by the requester of the
+/// timer. As this participates in intrusive data structures, it must be pinned
+/// before polling.
+#[derive(Debug)]
+pub(super) struct TimerEntry<TS: TimeSource> {
```

Comment:
I couldn't see why this is being made generic. I only saw a single time source.

Maybe this is legacy abstraction from the earlier timer that also needed to support `DelayQueue`?

Reply:
Initially I thought I'd need this for the loom stuff but indeed it turned out to be not as useful as I had thought. I'll pull that out.